src/core/cluster/quorum.rs

```diff
@@ -14,99 +16,187 @@ impl Quorum {
+                        key, local_put_ok, remote_ok_count, self.quorum
+                    );
+                    Err(err)
+                } else {
+                    let err = errors.remove(errors.len() - 1).into_inner();
+                    Err(err)
+                }
+            } else {
+                warn!("PUT[{}] succeed, but some data get into alien", key);
+                Ok(())
             }
-            let target_indexes = target_nodes.iter().map(Node::index);
-            let mut sup_nodes =
-                get_support_nodes(&self.mapper, target_indexes, additional_remote_writes)?;
-            debug!("PUT[{}] sup put nodes: {:?}", key, &sup_nodes);
-
-            let mut queries = Vec::new();
+        }
+    }
 
-            if let Err(op) = local_put {
-                let item = sup_nodes.remove(sup_nodes.len() - 1);
-                queries.push((item, op));
+    async fn background_put(
+        self,
+        mut rest_tasks: FuturesUnordered<JoinHandle<Result<NodeOutput<()>, NodeOutput<Error>>>>,
+        key: BobKey,
+        data: BobData,
+        mut failed_nodes: Vec<String>,
+    ) {
+        debug!("PUT[{}] ~~~BACKGROUND PUT TO REMOTE NODES~~~", key);
+        while let Some(join_res) = rest_tasks.next().await {
+            match join_res {
+                Ok(res) => match res {
+                    Ok(n) => debug!(
+                        "PUT[{}] successful background put to: {}",
+                        key,
+                        n.node_name()
+                    ),
+                    Err(e) => {
+                        error!("{:?}", e);
+                        failed_nodes.push(e.node_name().to_string());
+                    }
+                },
+                Err(e) => error!("{:?}", e),
             }
+        }
+        debug!("PUT[{}] ~~~PUT TO REMOTE NODES ALIEN~~~", key);
+        if let Err(e) = self.put_aliens(failed_nodes, key, data).await {
+            error!("{}", e);
+        }
+    }
+
+    pub(crate) async fn put_remote_nodes(
+        &self,
+        key: BobKey,
+        data: BobData,
+        at_least: usize,
+    ) -> (
+        FuturesUnordered<JoinHandle<Result<NodeOutput<()>, NodeOutput<Error>>>>,
+        Vec<NodeOutput<Error>>,
+    ) {
+        let local_node = self.mapper.local_node_name();
+        let target_nodes = get_target_nodes(&self.mapper, key)
+            .iter()
+            .filter(|node| node.name() != local_node);
+        put_at_least(key, data, target_nodes, at_least, PutOptions::new_local()).await
+    }
 
-            if additional_remote_writes > 0 {
-                let nodes = errors
+    pub(crate) async fn put_aliens(
+        &self,
+        mut failed_nodes: Vec<String>,
+        key: BobKey,
+        data: BobData,
+    ) -> Result<(), Error> {
+        let vdisk_id = self.mapper.id_from_key(key);
+        let operation = Operation::new_alien(vdisk_id);
+        let local_put = put_local_all(
```

Comment:
do we put at least one copy locally?

Reply:
It surprised me, but yes. Maybe this place needs a rework.

Reply:
#95